### ISSUE TITLE
Optimistic model with PromiseKit

### ISSUE THREAD

**Issue by zdenektopic** [mxcl/PromiseKit#257]
Hi,
I have just read this article https://medium.com/stanfy-engineering-practices/do-not-let-your-user-see-spinners-35b824c3ce2f on Medium.
I was wondering how could we implement such optimistic model with promises?
I don't think it is possible just like that if I want to use PromiseKit.

**Comment by nathanhosselton** [comment]
The example they gave doesn't map directly into a promises system because they are using a callback twice. And they are using a callback twice because they are combining a synchronous task with an asynchronous one. This wouldn't make sense in a promises system, which is intended to abstract asynchronicity, not fake it. Of course, you could still do it. One example: https://gist.github.com/nathanhosselton/ab23e210ce4c6cab79b1

Back in the context of the caller you could then use `when` on realPromise to execute additional code when realPromise resolves. This would achieve the goal of having the UI respond to both the fake request and the real one (i.e. mimicking using a callback twice) with an advantage over the article's example in that the responses would have context and could be discrete. And there are likely plenty of other ways the same effect could be reached; it just depends on the situation and how you've configured your API manager.

I want to state however that I think this example smells weird, not least because I think it betrays the purpose of promises. Instead, I would look more closely at the context of the request and determine whether the tasks (the fake request and the real request) should be combined and/or coupled to their UI responses in the first place, or whether a fake request is even necessary. My guess would be that there is probably a cleaner way.

**Comment by mxcl** [comment]
Yeah I'd go with what Nathan was saying. More elegant in Swift:

```swift
func follow() -> (Person, Promise<Person>) {
    return (person, Promise { … })
}

let (person, promise) = follow()
func setUI(person: Person) {
    //…
}
setUI(person)
promise.then(setUI)
promise.catch { error in
    // error
}
```

**Comment by jefferythomas** [comment]
While not perfectly analogous, this feels like [Cache & Fetch](http://promisekit.org/cache-n-fetch/).

    - (PMKPromise *)follow {
        return [Promise new:^(PMKPromiseFulfiller fulfill, PMKPromiseRejecter reject) {
            PMKPromise *real = [self.requestManager followPerson:self.person];
            Person *fake = [self.person copy];
            //…
            fulfill(PMKManifold(fake, real));
        }];
    }

    - (void)go {
        self.fetch.then(^(Person *fake, PMKPromise *real) {
            [self updateWithPerson:fake];
            return real;
        }).then(^(Person *real){
            [self updateWithPerson:real];
        });
    }

**Comment by nathanhosselton** [comment]
I agree, @jefferythomas, that system seems to work quite nicely here. Good catch.

**Comment by zdenektopic** [comment]
I see, it seems to be close enough, however, it still has 2 thens. Same thens actually. It gives you option to skip the "cached" data UI update, however it would be nice to make this work with single then, something like calling the fulfiller twice I guess?

**Comment by mxcl** [comment]
You can’t fulfill a promise twice, this is a *feature* generally. Just not for this pattern.

**Comment by zdenektopic** [comment]
what about something like startWith like in https://github.com/ReactiveX/RxSwift/?

**Comment by nathanhosselton** [comment]
I don't see an obvious example of `startWith`'s functionality. Can you summarize?
